### ISSUE TITLE
Cached errors are incorrectly rised for file checked with newer TS before

### ISSUE THREAD

**Issue by aczekajski** [microsoft/TypeScript#56710]
### 🔎 Search Terms

"errors from different version", "cached errors when changing TS version", "cached errors from newer tsserver", "older TS incorrectly shows errors rised by VSCode running newer TS"

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because it occurs when switching between TS versions

### ⏯ Playground Link

https://www.typescriptlang.org/play?noImplicitAny=false&ts=4.8.4#code/MYGwhgzhAECCA8AVAfNA3gWAFDV9AHgBQCU62eF0ATgKYAuArlQHbR0AWAljJHPGgQBc0CHSqdmAc2gBfZAG5yeGdiW4AnoXXDEpTFjWzsKg1mY0A7nBIA6IsRuaBFqgHspwgAyzi86AHp-Ni4YbmgwVhoqNypoCWgAVhsAJhToACMGOjiYZldsiOgomLjWABYbAA4bMqA

### 💻 Code

```ts
class A<T> {
    x() {
        return this as A<{ x: string }>;
    }

    y(y: T) {

    }
}

new A().x().y({ wrong: 0 }); // this is expected to be an error in 5.2.2 and expected to not be an error in 4.8.4
```

### 🙁 Actual behavior

First of all, the important thing is that this example code is correct in 4.8.4 and correctly rises an error in 5.2.2. When you open a project with this code that has a TS 4.8.4 installed, using VSCode with TypeScript version set to internal TS (5.2.2 in my case) instead of workspace TS (4.8.4), editor correctly shows an error in this file. However, when you run `node_modules/.bin/tsc --noEmit` from the terminal, which runs TS 4.8.4, it still outputs the same error, even though it shouldn't.

I made sure that TSC version 4.8.4 is being run in terminal using `node_modules/.bin/tsc -v`.

The error should not exist in 4.8.4 as this code is valid in 4.8.4 (with `noImplicitAny` set to `false`). No matter how many times you rerun `tsc` (4.8.4) is fails with errors cached from `tsserver` (5.2.2, run by VSCode) unless you change the file. 

Even if you now change the TS version used by VSCode to the workspace (4.8.4), the errors to that file are still being shown when running `tsc` without changing the file in question.

This behavior was noticed by several developers on MacOS and Windows.

### 🙂 Expected behavior

Typechecking code with tsc 4.8.4 should not raise errors cached from tsserver 5.2.2 that checked that file before and rised errors.

Playground links showing that the same code is:
- [correctly not an error on 4.8.4](https://www.typescriptlang.org/play?noImplicitAny=false&ts=4.8.4#code/MYGwhgzhAECCA8AVAfNA3gWAFDV9AHgBQCU62eF0ATgKYAuArlQHbR0AWAljJHPGgQBc0CHSqdmAc2gBfZAG5yeGdiW4AnoXXDEpTFjWzsKg1mY0A7nBIA6IsRuaBFqgHspwgAyzi86AHp-Ni4YbmgwVhoqNypoCWgAVhsAJhToACMGOjiYZldsiOgomLjWABYbAA4bMqA)
- [correctly an error on 5.2.2](https://www.typescriptlang.org/play?noImplicitAny=false&ts=5.2.2#code/MYGwhgzhAECCA8AVAfNA3gWAFDV9AHgBQCU62eF0ATgKYAuArlQHbR0AWAljJHPGgQBc0CHSqdmAc2gBfZAG5yeGdiW4AnoXXDEpTFjWzsKg1mY0A7nBIA6IsRuaBFqgHspwgAyzi86AHp-Ni4YbmgwVhoqNypoCWgAVhsAJhToACMGOjiYZldsiOgomLjWABYbAA4bMqA)

**Comment by RyanCavanaugh** [comment]
TS Server doesn't write anything to disk, or at least shouldn't, so this bug as described doesn't sound physically possible.

Can you share a repo that demonstrates this behavior?

**Comment by craigphicks** [comment]
Does the problem disappear if you run `npm i` (or equivalent)?

**Comment by aczekajski** [comment]
@RyanCavanaugh Oh, so the reason must be entirely different than I thought, sorry for the confusion ;o But the fact is that tsc 5.2.2 was never run by that user, only tsserver was run in 5.2.2, yet tsc 4.8.4 was raising errors exclusive to typescript 5.2.2, that why I thought it's definitely some cache.

I cannot share the original repo this happened in so I'll try preparing a small one with this issue isolated. However, the additional problem is that it's inconsistent, several devs saw it (both Mac and Win), but for example on my own PC it didn't happen at all. But I'll try anyway and we'll see, the issue is super strange.

I think we didn't try it but I'll ask again. We only made sure that there's only TS 4.8.4 installed in repo by running `npm list typescript`.
